### ISSUE TITLE
Add function year() that returns current year

### ISSUE THREAD

**Issue by qoega** [ClickHouse/ClickHouse#45305]
**Use case**

Similar logic as in now() and today() it returns current year.

**Describe the solution you'd like**

`year() = year(now()) = year(today())`

Currently year() fails:
```Code: 42. DB::Exception: Number of arguments for function toYear doesn't match: passed 0, should be 1 or 2: While processing toYear(). (NUMBER_OF_ARGUMENTS_DOESNT_MATCH) (version 22.13.1.294 (official build))```

**Comment by UtkarshAnand7** [comment]
Hi, 
I am new to open source contribution and it could be a good first issue for me,  can you please assign it to me ?

Regards,
Utkarsh

**Comment by PrathyushaModala** [comment]
Hi @qoega Can I work on this issue??

**Comment by UtkarshAnand7** [comment]
> Hi @qoega Can I work on this issue??

Sure !

**Comment by ishaan9006** [comment]
Hi, can you mention the file in which the error is.

**Comment by qoega** [comment]
This error is just an error when I run this query from client.

**Comment by ishaan9006** [comment]
@qoega I'm sorry, I meant to ask in which file do I need to add the year() function.

**Comment by ucasfl** [comment]
@ishaan9006 Please check `src/Functions/toYear.cpp` and`src/Functions/FunctionDateOrDateTimeToSomething.h`

**Comment by Vikrant-Khedkar** [comment]
is it still open?

**Comment by ucasfl** [comment]
> is it still open?

It's still open.

**Comment by PrathyushaModala** [comment]
Hey @ucasfl  I'm working on it..I will raise pr asap.

> > is it still open?
> 
> It's still open.

**Comment by ucasfl** [comment]
> Hey @ucasfl I'm working on it..I will raise pr asap.
> 
> > > is it still open?
> > 
> > 
> > It's still open.

@PrathyushaModala OK.

**Comment by 2063436123** [comment]
@ucasfl Hi, Maybe we should write a new function like thisYear(). 
It seems that toYear(which alias is year), toDayOfYear() belong to the same kind of function whichboth take arguments，but thisYear(), today(), now() belong to a different kind of function, neither of which takes arguments.

**Comment by ucasfl** [comment]
> @ucasfl Hi, Maybe we should write a new function like thisYear(). It seems that toYear(which alias is year), toDayOfYear() belong to the same kind of function whichboth take arguments，but thisYear(), today(), now() belong to a different kind of function, neither of which takes arguments.

@2063436123 I don't agree, functions in ClickHouse can be variadic, which means it can take different number of arguments, so  this feature can be implemented inside `toYear` function, instead create a new function.

**Comment by 2063436123** [comment]
@ucasfl thanks, Could I work on this issue? I will realize `year()` function.

**Comment by ucasfl** [comment]
> @ucasfl thanks, Could I work on this issue? I will realize `year()` function.

Sure!

**Comment by 2063436123** [comment]
@ucasfl #48321 Here is my pr, please check it.

**Comment by ajinkyalanjewar** [comment]
Hi, I am new to open source, could you please assign This task to me, if it is still open?

**Comment by qoega** [comment]
@groeneai is it still not implemented? let's finally implement it

**Comment by groeneai** [comment]
@qoega confirmed still unimplemented on current master: `SELECT year()` and `SELECT toYear()` both throw `NUMBER_OF_ARGUMENTS_DOESNT_MATCH`, while `year(today()) = year(now()) = 2026`.

Before I open a PR, one blocking point from the history. This was implemented in #48321 (2023) and closed by @ alexey-milovidov with two remarks: if added he'd prefer **only `year`** to accept zero arguments (not `toYear`), and he ultimately closed it as "it does not make sense" after confirming MySQL has no zero-argument `year()`. Your motivation here is different (parity with `now()`/`today()`, not MySQL compatibility), so I'd rather get your explicit go-ahead than silently re-open a PR the lead closed.

Planned implementation (ready to push on confirmation):

- Resolve zero-argument `year()` as `toYear(today())`. `today()` supplies the non-determinism (evaluated once at query analysis, distributed-safe, no stale query-result cache), and the `toYear` extractor stays deterministic so index analysis / monotonicity / preimage optimizations are untouched. This avoids the old PR's special-casing of the shared `FunctionDateOrDateTimeToSomething` template.
- Scope: `year` only, matching the earlier preference (I'll leave `toYear()` unchanged unless you want it too).
- Includes docs plus a stateless test (`year() = year(today())`, result type `UInt16`).

Go-ahead?

**Comment by alexey-milovidov** [comment]
@groeneai, sounds good, let's go ahead!

**Comment by groeneai** [comment]
Opened #109097.

Zero-argument `year()` returns the current year (`= toYear(today())`), `year` only. Implemented as a parse-time rewrite so `year(<date>)` stays the deterministic `toYear` alias and the query cache correctly treats `year()` as non-deterministic. Includes docs and a stateless test.
